cylc/flow/scheduler.py

```diff
@@ -408,6 +400,10 @@ async def initialise(self):
 
         self.profiler = Profiler(self, self.options.profile_mode)
 
+        # Reset graph node parser singleton
+        # (mainly to prevent state leakage between integration tests):
+        GraphNodeParser.get_inst().clear()
```

Comment:
> **mainly** to prevent state leakage between integration tests

Do you think this could cause any issues?

Reply:
If this only matters to integration tests, then it would make more sense to do it in the integration test fixtures (where it is only run by integration tests).

Otherwise, if it makes sense to do this for non-integration purposes, it would probably make more sense to do it in `cylc.flow.config` (where the `GraphNodeParser` is actually used) rather than in the scheduler code?

Reply:
Should this be done at the start of `WorkflowConfig.load_graph()` do you think?

Reply:
That sounds reasonable.

I don't think I had realised that there is a singleton here (seems a bit worrying for tests in general, not just integration ones). I wonder if we actually need it.

Reply:
Taking a look at the code, the `GraphNodeParser` appears to cache offsets (e.g. `-P1D`) and nodes (e.g. `foo[-P1D]`).

If my understanding of this is correct (it might not be), then it should be fine to preserve this cache within a workflow and even between workflows (`foo[-P1D]` means the same thing whatever workflow it is in). If so, it doesn't really make sense to clear this cache?

This commit added the `clear()` into the `get_graph_raw` method (https://github.com/cylc/cylc-flow/commit/8972b8eeb44ecce49c3d222ea444fc358be886ae) for performance reasons (which coming to think of it doesn't really make sense as the exact same nodes will have to be re-parsed in the next call, mb, I reviewed it).

Reply:
> foo[-P1D] means the same thing whatever workflow it is in

Not if one workflow is in integer cycling mode, hence the test failures

Reply:
> Not if one workflow is in integer cycling mode, hence the test failures

`Foo[-P1D]` has the exact same meaning whether the workflow is in integer, 365 or 360 day cycling mode, i.e. the task `foo` with an offset of `-P1D`. You won't find an offset of `P1D` in an integer cycling workflow, you'll find offsets like `-P1` instead, but having datetime entries in the cache shouldn't be a problem as they will never be retrieved.

Reply:
I feel like doing it here in `Scheduler.initialise()` which happens before loading the workflow config is safest for a 8.3.x PR?

Reply:
> I've pasted the traceback in the OP

Cheers.

Here's what the GraphNodeParser is caching about a node:

```
  123             self._nodes[node] = ( 
  124                 name, offset, TaskTrigger.standardise_name(output),
  125                 offset_is_from_icp, offset_is_irregular, offset_is_absolute)
```

I wouldn't expect any of this to clash between integer and datetime cycling modes e.g:

```
foo[-P1D] -> ('foo', ISO8601Duration('-P1D'), 'succeeded', False, False, False)
foo[-P1] -> ('foo', IntegerDuration('-P1'), 'succeeded', False, False, False)
```

It it does clash (which it evidently does), there are two possibilities, zero-duration offsets being stored with their respective class rather than `None`:

```
foo -> ('foo', IntegerDuration('-P0'), 'succeeded', False, False, False)
Or
foo -> ('foo', ISO8601Duration('-P0'), 'succeeded', False, False, False)
```

Or absolute dependencies being stored with their exact cycle point as the offset (in the respective class):

```
foo[^] -> ('foo', ISO8601Duration('<icp>'), 'succeeded', False, False, True)
Or
foo[^] -> ('foo', IntegerDuration('<icp>'), 'succeeded', False, False, True)
```

Reply:
If that reading is correct:

* This could never affect a real workflow, only test workflows.
* This could happen at validate time (because the graph nodes are also parsed then).

So ideally, we would have a fix that only impacts the tests (integration or otherwise) and works both for integration tests which actually start workflows AND any tests that load the config (e.g. validate tests) or poke at the GraphNodeParser.

Reply:
Ah I've just noticed we have an autouse fixture for resetting global state between tests

Reply:
Cheers.